src/tools/interop/idt/capture/platform/android/android.py

```diff
@@ -0,0 +1,220 @@
+                    print(f"Device {device_id} is offline, trying disconnect!")
+                    BashRunner(
+                        disconnect_command,
+                        sync=True,
+                        capture_output=False)
+                else:
+                    devices_auth[device_id] = device_is_auth
+            header_done = True
+        self.adb_devices = devices_auth
+        return self.adb_devices
+
+    def _only_one_device_connected(self) -> bool:
+        return len(self.adb_devices) == 1
+
+    def _get_first_connected_device(self) -> str:
+        return list(self.adb_devices.keys())[0]
+
+    def _set_device_if_only_one_connected(self) -> None:
+        if self._only_one_device_connected():
+            self.device_id = self._get_first_connected_device()
+            print(f'Only one device detected; using {self.device_id}')
+
+    def _log_adb_devices(self) -> None:
+        for dev in self.adb_devices:
+            print(dev)
+
+    @staticmethod
+    def _is_connection_str(adb_input_str: str) -> bool:
+        valid_ipv4 = False
+        port_entered = False
+        valid_port = False
+        split_on_colon = adb_input_str.split(":")
+        try:
+            ipaddress.IPv4Network(split_on_colon[0])
+            valid_ipv4 = True
+        except ValueError:
+            pass
+        if len(split_on_colon) > 1:
+            port_entered = True
+            try:
+                port = int(split_on_colon[1])
+                valid_port = port < 65535
+            except ValueError:
+                pass
+        valid_ip_no_port = valid_ipv4 and not port_entered
+        valid_ip_valid_port = valid_ipv4 and valid_port
+        return valid_ip_no_port or valid_ip_valid_port
+
+    def _check_connect_wireless_adb(self, temp_device_id: str) -> None:
+        # TODO: Wireless debugging does not work from container
+        if Android._is_connection_str(temp_device_id):
+            connect_command = f"adb connect {temp_device_id}"
+            print(
+                f"Detected connection string; attempting to connect: {connect_command}")
+            BashRunner(connect_command, sync=True, capture_output=False)
+            self.get_adb_devices()
+
+    def _device_id_user_input(self) -> None:
+        print('If there is no output below, press enter after connecting your phone under test OR')
+        print('Enter (copy paste) the target device id from the list of available devices below OR')
+        print('Enter $IP4:$PORT to connect wireless debugging.')
+        self._log_adb_devices()
+        temp_device_id = input('').strip()
+        self._check_connect_wireless_adb(temp_device_id)
+        if self._only_one_device_connected():
+            self._set_device_if_only_one_connected()
+        elif temp_device_id not in self.adb_devices:
+            print('Entered device not in adb devices!')
+        else:
+            self.device_id = temp_device_id
+
+    def _choose_device_id(self) -> None:
```

Comment:
Here is a snippet for a nice urwid based selection prompt.

```py
def prompt_user_for_selection(options: Iterable[str]) -> Optional[str]:
    selection = None

    def item_chosen(choice, *_args):
        nonlocal selection
        selection = choice

        raise urwid.ExitMainLoop()

    body = [urwid.Text('Select between the available ports...'), urwid.Divider()]
    for c in options:
        button = urwid.Button(c)
        urwid.connect_signal(button, 'click', item_chosen, user_args=[c])
        body.append(urwid.AttrMap(button, None, focus_map='reversed'))

    menu = urwid.ListBox(urwid.SimpleFocusListWalker(body))
    urwid.MainLoop(menu, palette=[('reversed', 'standout', '')]).run()

    return selection
```

Reply:
This is a neat library, it will be cool to work this in eventually.